### ISSUE TITLE
control pytorch's massive initial memory overhead GPU

### ISSUE THREAD

**Issue by xxradon** [pytorch/pytorch#31881]

**Comment by soumith** [comment]
the original discussion captures the issue well.

There isn't a way to control the memory, other than recompiling pytorch from scratch with only a subset of functionality. This is a CUDA limitation.

One possible thing to debate is to see if we can have a compile-time flag that allows finer grained disabling of types, for example not compiling particular {device/dtypes} such as {cuda, float64} and {cuda, int16} -- which aren't useful on a Jetson Nano.

cc: @gchanan

**Comment by TinDang97** [comment]
I take a lot of issues with it. Like I can't run multi-node on docker base on Kafka to speed up my project cause lost massive memory. It same as a higher cost. Now, I only run multiprocess in a single model.

**Comment by chaitanyarit** [comment]
One way to solve this in linux is to limit the memory to a process by using cgroups. 

# install cgroups:
sudo apt-get install cgroup-bin cgroup-lite cgroup-tools cgroupfs-mount libcgroup1
# create group 
sudo cgcreate -t $USER:$USER -a $USER:$USER  -g memory:myGroup
# set memory limits
echo 500M > /sys/fs/cgroup/memory/myGroup/memory.limit_in_bytes   # restrict memory
echo 5G > /sys/fs/cgroup/memory/myGroup/memory.memsw.limit_in_bytes  # restrict swap space
# edit /etc/default/grub add the below line 
GRUB_CMDLINE_LINUX_DEFAULT="cgroup_enable=memory swapaccount=1"
# sudo update grub
set -e
exec grub-mkconfig -o /boot/grub/grub.cfg "$@"
# reboot and run process
cgexec -g memory:myGroup python infer.py # any process 

This will force the pytorch to limit the allocation of memory freely. Loading time might increase but the initial memory overhead is reduced. On Jetson TX2 , it reduced the memory consumption for mobilenet-ssd from 25% to 5% when I restricted the memory to 300M.

**Comment by lamhoangtung** [comment]
Any update on this guys ?

If Pytorch would take a tons of effort to reduce the CUDA runtime memory consumption then how much effort would it take to make Scripted Model run with less CUDA runtime memory ?

**Comment by UlrichFreitag** [comment]
Yeah that feature would be very useful. The thing is, we can see an initial memory footprint increasing (by PyTorch) between each new GPU architecture... Would that not be possible to dynamically load the necessary kernels when needed?

**Comment by ezyang** [comment]
@malfet you were talking to me about this very issue recently right? I suggested that you should look into mobile selective build, which does exactly what people are asking for here: a way to only compile kernels in PyTorch that you need. This seems like exactly the issue being reported here.

One thing that I'm a little confusing is @UlrichFreitag, you seem to imply that the memory footprint increase is due to new GPU architectures. Is this actually the case? If so, it seems like recompiling PyTorch with only the arch you need should help. But that would be very surprising to me: I would hope CUDA would avoid loading kernels for an arch which is not useful.

**Comment by lamhoangtung** [comment]
@ezyang So how can I compile only needed kernel for libtorch? I spent quite the effort to move down to libtorch to hope for a better production deployment but GPU memory footprint doest not improve at all ;(

**Comment by ezyang** [comment]
also cc @ljk53

**Comment by hyperfraise** [comment]
I'm very interested in this too

**Comment by xsacha** [comment]
+1 this is a blocker for usage on smaller devices like Jetson Nano and the only reason we require the NX instead.
